src/aosm/azext_aosm/generate_nsd/nf_ret.py

```diff
@@ -0,0 +1,165 @@
+# --------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT
+# License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------
+"""Handles the creation of a resource element template for a network function."""
+
+import json
+
+from typing import Dict, Any, List, Union
+from knack.log import get_logger
+
+from azext_aosm._configuration import NFDRETConfiguration
+from azext_aosm.util.constants import CNF, VNF
+from azext_aosm.util.management_clients import ApiClients
+from azext_aosm.vendored_sdks.models import NetworkFunctionDefinitionVersion, NFVIType
+
+
+logger = get_logger(__name__)
+
+
+class nfRET:
+    """
+    Represents a single network function resource element template withing an NSD.
+    """
+
+    def __init__(
+        self, api_clients: ApiClients, config: NFDRETConfiguration, cg_schema_name: str
+    ) -> None:
+        self.config = config
+        self.cg_schema_name = cg_schema_name
+        nfdv = self._get_nfdv(config, api_clients)
+        print(
+            f"Finding the deploy parameters for {self.config.name}:{self.config.version}"
+        )
+
+        if not nfdv.deploy_parameters:
+            raise NotImplementedError(
+                f"NFDV {self.config.name} has no deploy parameters, cannot generate NSD."
+            )
+        self.deploy_parameters: Dict[str, Any] = json.loads(nfdv.deploy_parameters)
+
+        self.nf_type = self.config.name.replace("-", "_")
+        self.nfdv_parameter_name = f"{self.nf_type}_nfd_version"
+        self.config_mapping_filename = f"{self.config.name}_config_mapping.json"
+
+    def _get_nfdv(
+        self, config: NFDRETConfiguration, api_clients
+    ) -> NetworkFunctionDefinitionVersion:
+        """Get the existing NFDV resource object."""
+        print(
+            "Reading existing NFDV resource object "
+            f"{config.version} from group {config.name}"
+        )
+        nfdv_object = api_clients.aosm_client.network_function_definition_versions.get(
+            resource_group_name=config.publisher_resource_group,
```

Comment:
Should we not be using proxy publishers for this?  Appreciate that this isn't a change you made.  Also, I haven't finished watching Jen's recording on proxy publishers, but I thought we weren't supposed to know the publisher RG any more, and just get the publisher from location and name?

Reply:
I've raised https://dev.azure.com/msazuredev/AzureForOperators/_workitems/edit/849121 to cover this.